// hoyoapi/dist/esm/index.mjs
import {
  DiaryEnum,
  DiaryMonthEnum,
  GenshinDiaryModule,
  GenshinImpact,
  GenshinRecordModule,
  GenshinRegion,
  GenshinTCGModule,
  SpiralAbyssScheduleEnum,
  getGenshinRegion
} from "./chunk-XFOQMEZ2.mjs";
import {
  HIRecordModule,
  HonkaiImpact,
  HonkaiRegion,
  getHi3Region
} from "./chunk-QXLLCPQD.mjs";
import {
  ForgottenHallScheduleEnum,
  HSRRecordModule,
  HonkaiStarRail,
  HsrRegion,
  getHsrRegion
} from "./chunk-E4N6QNCC.mjs";
import {
  DailyModule,
  RedeemModule
} from "./chunk-KCSC7RXQ.mjs";
import {
  ACCOUNT_API,
  BBS_API,
  Cookie,
  DAILY_CLAIM_API,
  DAILY_INFO_API,
  DAILY_REWARD_API,
  DEFAULT_REFERER,
  GAME_RECORD_CARD_API,
  GENSHIN_DIARY_DETAIL_API,
  GENSHIN_DIARY_LIST_API,
  GENSHIN_RECORD_AVATAR_BASIC_INFO_API,
  GENSHIN_RECORD_CHARACTER_API,
  GENSHIN_RECORD_DAILY_NOTE_API,
  GENSHIN_RECORD_INDEX_API,
  GENSHIN_RECORD_SPIRAL_ABYSS_API,
  GENSHIN_TCG_BASICINFO,
  GENSHIN_TCG_CARDLIST,
  GENSHIN_TCG_CHALLANGE_DECK,
  GENSHIN_TCG_CHALLANGE_RECORD,
  GENSHIN_TCG_CHALLANGE_SCHEDULE,
  GENSHIN_TCG_MATCHLIST,
  GamesEnum,
  HI_RECORD_ABYSS_API,
  HI_RECORD_ARENA_API,
  HI_RECORD_CHARACTER_API,
  HI_RECORD_ELYSIAN_API,
  HI_RECORD_INDEX_API,
  HK4E_API,
  HSR_RECORD_CHARACTER_API,
  HSR_RECORD_FORGOTTEN_HALL_API,
  HSR_RECORD_INDEX_API,
  HSR_RECORD_NOTE_API,
  HTTPRequest,
  HoyoAPIError,
  Hoyolab,
  Language,
  LanguageEnum,
  PUBLIC_API,
  REDEEM_CLAIM_API,
  USER_GAMES_LIST
} from "./chunk-VUVEAPAF.mjs";
export {
  ACCOUNT_API,
  BBS_API,
  Cookie,
  DAILY_CLAIM_API,
  DAILY_INFO_API,
  DAILY_REWARD_API,
  DEFAULT_REFERER,
  DailyModule,
  DiaryEnum,
  DiaryMonthEnum,
  ForgottenHallScheduleEnum,
  GAME_RECORD_CARD_API,
  GENSHIN_DIARY_DETAIL_API,
  GENSHIN_DIARY_LIST_API,
  GENSHIN_RECORD_AVATAR_BASIC_INFO_API,
  GENSHIN_RECORD_CHARACTER_API,
  GENSHIN_RECORD_DAILY_NOTE_API,
  GENSHIN_RECORD_INDEX_API,
  GENSHIN_RECORD_SPIRAL_ABYSS_API,
  GENSHIN_TCG_BASICINFO,
  GENSHIN_TCG_CARDLIST,
  GENSHIN_TCG_CHALLANGE_DECK,
  GENSHIN_TCG_CHALLANGE_RECORD,
  GENSHIN_TCG_CHALLANGE_SCHEDULE,
  GENSHIN_TCG_MATCHLIST,
  GamesEnum,
  GenshinDiaryModule,
  GenshinImpact,
  GenshinRecordModule,
  GenshinRegion,
  GenshinTCGModule,
  HIRecordModule,
  HI_RECORD_ABYSS_API,
  HI_RECORD_ARENA_API,
  HI_RECORD_CHARACTER_API,
  HI_RECORD_ELYSIAN_API,
  HI_RECORD_INDEX_API,
  HK4E_API,
  HSRRecordModule,
  HSR_RECORD_CHARACTER_API,
  HSR_RECORD_FORGOTTEN_HALL_API,
  HSR_RECORD_INDEX_API,
  HSR_RECORD_NOTE_API,
  HTTPRequest,
  HonkaiImpact,
  HonkaiRegion,
  HonkaiStarRail,
  HoyoAPIError,
  Hoyolab,
  HsrRegion,
  Language,
  LanguageEnum,
  PUBLIC_API,
  REDEEM_CLAIM_API,
  RedeemModule,
  SpiralAbyssScheduleEnum,
  USER_GAMES_LIST,
  getGenshinRegion,
  getHi3Region,
  getHsrRegion
};
